codegen/generate.go

```diff
@@ -14,3 +39,176 @@ func GenerateCode(data *Data) error {
 		Packages:        data.Config.Packages,
 	})
 }
+
+func generatePerSchema(data *Data) error {
+	err := generateRootFile(data)
+	if err != nil {
+		return err
+	}
+
+	builds := map[string]*Data{}
+
+	err = addObjects(data, &builds)
+	if err != nil {
+		return err
+	}
+
+	err = addInputs(data, &builds)
+	if err != nil {
+		return err
+	}
+
+	err = addInterfaces(data, &builds)
+	if err != nil {
+		return err
+	}
+
+	err = addReferencedTypes(data, &builds)
+	if err != nil {
+		return err
+	}
+
+	for filename, build := range builds {
+		if filename == "" {
+			continue
+		}
+
+		dir := data.Config.Exec.DirName
+		path := filepath.Join(dir, filename)
+
+		err = templates.Render(templates.Options{
+			PackageName:     data.Config.Exec.Package,
+			Filename:        path,
+			Data:            build,
+			RegionTags:      true,
+			GeneratedHeader: true,
+			Packages:        data.Config.Packages,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func filename(p *ast.Position, config *config.Config) string {
+	name := "common!"
+	if p != nil && p.Src != nil {
+		gqlname := filepath.Base(p.Src.Name)
+		ext := filepath.Ext(p.Src.Name)
+		name = strings.TrimSuffix(gqlname, ext)
+	}
+
+	filenameTempl := config.Exec.FilenameTemplate
+	if filenameTempl == "" {
+		filenameTempl = "{name}.generated.go"
+	}
+
+	return strings.ReplaceAll(filenameTempl, "{name}", name)
+}
+
+func addBuild(filename string, p *ast.Position, data *Data, builds *map[string]*Data) {
+	buildConfig := *data.Config
+	if p != nil {
+		buildConfig.Sources = []*ast.Source{p.Src}
+	}
+
+	(*builds)[filename] = &Data{
+		Config:           &buildConfig,
+		QueryRoot:        data.QueryRoot,
+		MutationRoot:     data.MutationRoot,
+		SubscriptionRoot: data.SubscriptionRoot,
+		AllDirectives:    data.AllDirectives,
+	}
+}
+
+// Root file contains top-level definitions that should not be duplicated across the generated
+// files for each schema file.
+func generateRootFile(data *Data) error {
+	dir := data.Config.Exec.DirName
+	path := filepath.Join(dir, "root!.generated.go")
```

Comment:
We're seeing issues with our tooling due to the '!' in this filename. Any reason why it has to have an exclamation point in it?  Quoting with "" doesn't even help, so it can be tricky to deal with.
```
bash$ echo "!root"
bash: !root: event not found
bash$ echo '!root'
!root
bash$
```

Reply:
My original intention was to minimize potential naming conflicts in case people had schema files named `root.graphql` in their project, but no reason it needs to be an exclamation point- could be changed to another symbol or just accept the possibility of conflicts

Reply:
Thank you @kevinmbeaulieu - just sent a PR to change the name to `root_.generated.go`
https://github.com/99designs/gqlgen/pull/1701